### ISSUE TITLE
GraphQL Plugin Not Reporting Spans

### ISSUE THREAD

**Issue by icirellik** [DataDog/dd-trace-js#951]
**Describe the bug**

TLDR: I think I found the line causing the problem. Skip to Potential Resolution.

I was trying to instrument our application using the `dd-trace` `graphql` plugin and noticed that anytime it was enabled our entire applications api tracing would stop working. The application works exceptionally when only the `express` and `http` plugins are loaded. The configuration that caused tracing to stop working is below.:

```ts
import tracer from "dd-trace"

function init() {
  tracer.init({
    service: "azul",
    hostname: "localhost",
    plugins: false,
    logger: { debug, error },
  })
  tracer.use("express", {
    service: "azul",
    headers: ["User-Agent", "X-User-ID"],
  })
  tracer.use("http", {
    service: "azul.http-client",
  })
  tracer.use("graphql", {
    service: "azul.graphql",
  })
}
```

If the `graphql` plugin is removed it starts working again.

**Potential Resolution**

After digging into the issue I discovered what I believe that the root cause. When I comment out the following line tracing starts working again and includes all the `graphql` traces.

https://github.com/DataDog/dd-trace-js/blob/f1204494de67121145adc10dfa0ea463b50aa080/packages/datadog-plugin-graphql/src/index.js#L58

A few lines above, you can see that `startSpan` is called. I believe that this in combination with the early exit is the issue. When it exits early it leaves an open span that `finish` will not be called for. It wasn't obvious at first that this would be an issue. That comment on the line makes it appear that this would work as expected.

As I continued to dig I discovered that the [exporter](https://github.com/DataDog/dd-trace-js/blob/f1204494de67121145adc10dfa0ea463b50aa080/packages/dd-trace/src/exporters/agent/index.js#L18), the method that eventually sends the spans to the DataDog Agent would never be called. It was being blocked because the `started` and `finished` span length would never be equal once of the spans that were never finished.

https://github.com/DataDog/dd-trace-js/blob/f1204494de67121145adc10dfa0ea463b50aa080/packages/dd-trace/src/span_processor.js#L14

There are a couple potential solutions here and I'd be willing to PR if desired. There either needs to be a way to `abort` a span or just don't create the span until we know it will be needed.

**Related Issues**

I think this issue is related to the following:

* https://github.com/DataDog/dd-trace-js/issues/706

**Environment**

* **Operation system:** macOS, Linux
* **Node version:** 10.20.1
* **Tracer version:** 0.20.3
* **Agent version:** datadog/agent:latest

**Comment by rochdev** [comment]
Thanks for the thorough investigation!

> There either needs to be a way to abort a span or just don't create the span until we know it will be needed.

I'm actually not sure if the span should be aborted, or if it should always be finished but without the document tags when not available. Even if the returned document doesn't have an operation in its definition, the parsing still occurred and it might make sense to report this timing as a span. Said that, I think simply moving `finish` to a `finally` block could resolve the issue and guarantee that it's always called.

> There are a couple potential solutions here and I'd be willing to PR if desired.

I should be able to get this fixed pretty quickly if you prefer, but otherwise definitely feel free to open a PR :) Just jet me know so that we're not duplicating work.

**Comment by icirellik** [comment]
Moving it the `finally` block makes perfect sense. I wasn't sure under what conditions a span should start and end. Sounds like you'll be able to get it fixed quicker than me, I'll just leave it to you.

**Comment by meiamsome** [comment]
Hi, I was just experiencing a similar issue, in that case whenever `parse` was being used to parse a schema (or fragments without an `OperationDefinition` node) then this will fail. The PR #960 worked in that code base when applied as well. 

Of particular interest when using this in a standard GraphQL server, it would appear that sending valid GraphQL documents that are not Operations (i.e. just fragments). would result in hitting this condition, and so all such requests would not be correctly reported in DataDog.

**Comment by rochdev** [comment]
Fixed in 0.21.0. Thanks for the PR @icirellik !

**Comment by MadRabbit** [comment]
hmm, i think i'm having this one too. i'm on 0.24 and latest express-graphql + graphql. the only span i see is the `graphql.parse`, there is nothing underneath it

**Comment by xinghul** [comment]
Hello, I'm using `v1.5.1`, and I can only see `graphql.parse` from the resources:
<img width="2359" alt="Screen Shot 2022-02-17 at 3 16 56 PM" src="https://user-images.githubusercontent.com/6555861/154587755-adc7b75b-e7b3-4ce2-9b7b-b7fe386a9c8c.png">

Here's my setup just FYI:
```javascript
tracer.init({
  service: serviceName,
});

tracer.use('graphql', {
  service: `${serviceName}-graphql`,
});
```

**Comment by rochdev** [comment]
@xinghul This would usually be caused by the tracer being initialized after other modules like `graphql` have already been imported. Is the snippet you shared the very first thing happening in the application, before any other `require` or `import` statements? Also, is this something you are able to reproduce with code that you can share?

**Comment by gschier** [comment]
> Hello, I'm using `v1.5.1`, and I can only see `graphql.parse` from the resources: <img alt="Screen Shot 2022-02-17 at 3 16 56 PM" width="2000" src="https://user-images.githubusercontent.com/6555861/154587755-adc7b75b-e7b3-4ce2-9b7b-b7fe386a9c8c.png">
> 
> I'm using Apollo Server, not sure if that's related, here's my setup just FYI:
> 
> ```js
> tracer.init({
>   service: serviceName,
> });
> 
> tracer.use('graphql', {
>   service: `${serviceName}-graphql`,
> });
> ```

Hi there, we're having the same issue! Did you ever figure out the problem?

**Comment by gschier** [comment]
> Hello, I'm using `v1.5.1`, and I can only see `graphql.parse` from the resources: <img alt="Screen Shot 2022-02-17 at 3 16 56 PM" width="2000" src="https://user-images.githubusercontent.com/6555861/154587755-adc7b75b-e7b3-4ce2-9b7b-b7fe386a9c8c.png">
> 
> I'm using Apollo Server, not sure if that's related, here's my setup just FYI:
> 
> ```js
> tracer.init({
>   service: serviceName,
> });
> 
> tracer.use('graphql', {
>   service: `${serviceName}-graphql`,
> });
> ```

Hey there @xinghul, did you ever figure out the problem? We're having the same issue.

**Comment by gschier** [comment]
Ah, nope, it was my fault. I didn't realize our production entrypoint was different and just had to inject it as the first import over there.
